nav2_tasks/README.md

```diff
@@ -0,0 +1,108 @@
+
+This approach allows for easily replacing any task with an alternative implementation. The replacement module must simply implement the correct task interface. 
+
+One can define additional tasks using the TaskClient and TaskSever templates. To do so, one specifies the command message that the task is to receive as well as the result message (similar in spirit to ROS1's .action file). 
+
+For example, to define a new task, named *DoSomething*, that takes a String as input and an Empty message as a result, 
+
+```C++
+namespace nav2_tasks
+{
+
+using DoSomethingCommand = std_msgs::msg::String;
+using DoSomethingResult = std_msgs::msg::Empty;
+
+using DoSomethingTaskClient = TaskClient<DoSomethingCommand, DoSomethingResult>;
+using DoSomethingTaskServer = TaskServer<DoSomethingCommand, DoSomethingResult>;
+
+template<>
+inline const char * getTaskName<DoSomethingCommand, DoSomethingResult>()
+{
+  return "DoSomethingTask";
+}
+```
+
+**Note**: The getTaskName template is required (it's used to name the topics used by the implementation).
+
+Then, to implement the task server, a derived class inherits from this newly-defined task type and overrides the execute() method. For example,
+
+```C++
+class DoSomethingImplementation : public nav2_tasks::DoSomethingTaskServer
+{
+public:
+  DoSomethingImplementation();
+  ~DoSomethingImplementation();
+
+  nav2_tasks::TaskStatus execute(
+    const nav2_tasks::DoSomethingCommand::SharedPtr command) override;
+};
+
+TaskStatus
+DoSomethingImplementation::execute(const nav2_tasks::DoSomethingCommand::SharedPtr command)
+{  
+  for (;;) {
+    
+    // TODO: Do a bit of the task        
+
+    // Before we loop again to do more work, check if we've been canceled
+    if (cancelRequested()) {
+      RCLCPP_INFO(get_logger(), "DwaController::execute: task has been canceled");
+      setCanceled();
+      return TaskStatus::CANCELED;
+    }
+    
+    // If we've successfully completed the task, return the result
+    if (done) {
+      nav2_tasks::DoSomethingResult result; 
+      
+      // TODO: Set fields in the result message, if any
+      
+      setResult(result);
+      return TaskStatus::SUCCEEDED;
+    }
+  }  
+}
+```
+
+## Implementation
+
+The implementation of TaskClient and TaskServer is simplistic as the intention is to rely upon a ROS2 ActionLib (or equivalent) when it is available. 
+
+The current implementation uses four topics, one each for **Command**, **Cancel**, **Result**, and **Status** messages. The Command and Result messages are provided by the user of the classes, while the Result and Status messages are internal to the implementation. The Status message communicates whether a task has succeeded, failed, or has been canceled, while the Result message is used to return the result of the command, if successful. To keep things simple, the task library does not implement a feedback mechanism as the ROS1 ActionLib does. 
```

Comment:
Do you mean "The Command and _Cancel_ messages are provided..."? I think you forgot to mention the Cancel topic.